### PR TITLE
DependencyGraphBuilder to Respect Exclusion in Dependency

### LINKED ISSUE #799: DependencyGraphBuilder ignoring exclusions
image

Running `DependencyTreePrinter` with `com.google.cloud.tools:jib-maven-plugin:1.4.1-SNAPSHOT` fails:

```
com.google.cloud.tools:jib-maven-plugin:1.4.1-SNAPSHOT : Failed to retrieve dependency information:There were failure(s) in dependency resolution
com.google.cloud.tools:jib-maven-plugin:jar:1.4.1-SNAPSHOT (compile) / org.apache.maven:maven-plugin-api:jar:3.6.1 (compile) / org.eclipse.sisu:org.eclipse.sisu.plexus:jar:0.3.3 (compile) / javax.enterprise:cdi-api:jar:1.0 (compile) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile?) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile): org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 in central (http://repo1.maven.org/maven2/)
com.google.cloud.tools:jib-maven-plugin:jar:1.4.1-SNAPSHOT (compile) / org.apache.maven:maven-core:jar:3.6.1 (compile) / org.eclipse.sisu:org.eclipse.sisu.plexus:jar:0.3.3 (compile) / javax.enterprise:cdi-api:jar:1.0 (compile) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile?) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile): org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 in central (http://repo1.maven.org/maven2/)
com.google.cloud.tools:jib-maven-plugin:jar:1.4.1-SNAPSHOT (compile) / org.apache.maven:maven-core:jar:3.6.1 (compile) / org.apache.maven:maven-plugin-api:jar:3.6.1 (compile) / org.eclipse.sisu:org.eclipse.sisu.plexus:jar:0.3.3 (compile) / javax.enterprise:cdi-api:jar:1.0 (compile) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile?) / org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 (compile): org.eclipse.aether.resolution.DependencyResolutionException: Could not find artifact org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0 in central (http://repo1.maven.org/maven2/)
```

[`org.eclipse.sisu:org.eclipse.sisu.plexus:0.3.3`](https://search.maven.org/artifact/org.eclipse.sisu/org.eclipse.sisu.plexus/0.3.3/eclipse-plugin) dependency with exclusions:

```
    <dependency>
      <groupId>javax.enterprise</groupId>
      <artifactId>cdi-api</artifactId>
      <version>1.0</version>
      <exclusions>
        <exclusion>
          <groupId>javax.el</groupId>
          <artifactId>el-api</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.jboss.ejb3</groupId>
          <artifactId>jboss-ejb3-api</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.jboss.interceptor</groupId>
          <artifactId>jboss-interceptor-api</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

DependencyGraphBuilder does not take the exclusion element into account, and thus `DependencyTreePrinter` fails due to missing artifacts (`org.jboss.ejb3:jboss-ejb3-api:jar:3.1.0`).

This is not a problem in the enforcer rule because the enforcer rule does not use DependencyGraphBuilder. But problematic for #795 as I was thinking to use DependencyGraphBuilder to get the path information.

# Example: io.grpc:grpc-core

`io.grpc:grpc-core:1.13.1` excludes guava in opencensus-api. Should dashboard's linkage check and dependency convergence respect the exclusions?

image

### LINKED ISSUE #1056: Protobuf-lite errors in BOM
image

# protobuf-java-3.11.0.jar
2 target classes causing linkage errors referenced from 2 source classes.

(protobuf-lite-3.0.1.jar) com.google.protobuf.CodedOutputStream$NioEncoder's method writeMessage(int arg1, com.google.protobuf.MessageLite arg2, com.google.protobuf.Schema arg3) is not implemented in the class, referenced from 1 class ▶

(protobuf-lite-3.0.1.jar) com.google.protobuf.CodedOutputStream$NioEncoder's method writeMessageNoTag(com.google.protobuf.MessageLite arg1, com.google.protobuf.Schema arg2) is not implemented in the class, referenced from 1 class ▶

The following paths to the jar file from the BOM are found in the dependency tree:

com.google.protobuf:protobuf-java:3.11.0 (compile)
com.google.protobuf:protobuf-java-util:3.11.0 (compile) / com.google.protobuf:protobuf-java:3.11.0 (compile)

This error seems a genuine error.

image

# protobuf-lite (old. Latest: 3.0.1)

https://github.com/protocolbuffers/protobuf/blob/javalite/java/lite.md

# protobuf-javalite (new. Latest: 3.11.1)

https://github.com/protocolbuffers/protobuf/blob/master/java/lite.md

Should gRPC use protobuf-javalite?

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #799 .
Fixes #1056 . With this PR, The dashboard stops showing the protobuf-lite linkage errors.

DependencyGraphBuilder to use DependencyNode, which has exclusions information as well as Artifact.

# Difference in the Dashboard

The number of the upper bounds errors decreased. The number of artifacts failing to converge decreased.

image

## Difference in dependency convergence. Example in grpc-alts:
image

## New Linkage Errors

Because Netty's dependency on `com.oracle.substratevm:svm` excludes unnecessary artifacts ([link](https://search.maven.org/artifact/io.netty/netty-common/4.1.42.Final/jar)), the dashboard now shows these missing classes as linkage errors. 

image

I ran LinkageCheckerMain of this snapshot version against Apache Beam and did not see problem caused by exclusions.

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 1]
```diff
@@ -108,28 +108,32 @@ private static String osDetectedArch() {
...
-    Map<String, DependencyNode> cache =
+    // The cache key includes exclusion elements of Maven artifacts
+    Map<Dependency, DependencyNode> cache =
         includeProvidedScope ? cacheWithProvidedScope : cacheWithoutProvidedScope;
-    String cacheKey =
-        dependencyArtifacts.stream().map(Artifacts::toCoordinates).collect(Collectors.joining(","));
-    if (cache.containsKey(cacheKey)) {
-      return cache.get(cacheKey);
+    Dependency cacheKey = null;
+    if (dependencyNodes.size() == 1) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: In our usage (dashboard and linkage monitor), the cache for multiple `dependencyArtifacts` was not used.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java [hunk 2]
```diff
@@ -201,48 +201,4 @@ public void testArtifactToClasspath_notToGenerateRepositoryException()
     List<Path> paths = ClassPathBuilder.artifactsToClasspath(ImmutableList.of(jamonApiArtifact));
     Truth.assertThat(paths).isNotEmpty();
   }
-
-  @Test
-  public void testArtifactToClasspath_reportAggregatedRepositoryException()
-      throws RepositoryException {
-    // jamon has transitive dependency to jmxtools, which does not exist in Maven central
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java]: With the change, resolving `com.jamonapi:jamon:2.81` no longer throws an exception.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 3]
```diff
@@ -214,20 +214,31 @@ public void testLinkageReports() {
             "106 target classes causing linkage errors referenced from 516 source classes.");
 
     Nodes dependencyPaths = details.query("//p[@class='linkage-check-dependency-paths']");
-    Node dependencyPathMessageOnProblem = dependencyPaths.get(dependencyPaths.size() - 2);
+    Node dependencyPathMessageOnProblem = dependencyPaths.get(dependencyPaths.size() - 4);
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: The position of the error message has changed.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 4]
```diff
@@ -214,20 +214,31 @@ public void testLinkageReports() {
...
     Assert.assertEquals(
-        "The following path contains guava-jdk5-17.0.jar:",
+        "The following paths contain guava-jdk5-13.0.jar:",
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: With the change, the version of guava-jdk5 picked up by the dashboard has changed.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java [hunk 5]
```diff
@@ -42,34 +42,33 @@ public void testFindUpdates() throws RepositoryException {
 
     DependencyGraph graph = DependencyGraphBuilder.getCompleteDependencies(core);
     List<Update> updates = graph.findUpdates();
-    List<String> strings = updates.stream().map(e -> e.toString()).collect(Collectors.toList());
-
-    // ordering not working yet
-    // TODO get order working
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphIntegrationTest.java]: Fixed this TODO by `inOrder()`.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 6]
```diff
@@ -124,4 +125,22 @@ public void testSetDetectedOsSystemProperties_netty4Dependency() throws Reposito
     List<Artifact> artifacts = DependencyGraphBuilder.getDirectDependencies(nettyArtifact);
     Truth.assertThat(artifacts).isNotEmpty();
   }
+
+  @Test
+  public void testSetDetectedOsSystemProperties_grpcProtobufExclusion() throws RepositoryException {
+    // Grpc-protobuf depends on grpc-protobuf-lite with protobuf-lite exclusion.
+    // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1056
+    Artifact nettyArtifact = new DefaultArtifact("io.grpc:grpc-protobuf:1.25.0");
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: These coordinates don't point to netty.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 7]
```diff
@@ -124,4 +125,22 @@ public void testSetDetectedOsSystemProperties_netty4Dependency() throws Reposito
...
+
+    DependencyGraph dependencyGraph =
+        DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph(
+            ImmutableList.of(nettyArtifact));
+
+    Truth.assertThat(dependencyGraph.list())
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: This is hard to follow. Can it be spit out into several statements?

Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: Extracted Correspondence into a variable.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: Good catch. Fixed.
